contracts/sidechain/FuelMessagePortal.sol

```diff
@@ -0,0 +1,313 @@
+    // General Public Functions //
+    //////////////////////////////
+
+    /// @notice Used by message receiving contracts to get the address on Fuel that sent the message
+    /// @return sender the address of the sender on Fuel
+    function getMessageSender() external view returns (bytes32) {
+        require(s_incomingMessageSender != NULL_MESSAGE_SENDER, "Current message sender not set");
+        return s_incomingMessageSender;
+    }
+
+    /// @notice Gets the number of decimals used in the Fuel base asset
+    /// @return decimals of the Fuel base asset
+    function getFuelBaseAssetDecimals() public pure returns (uint8) {
+        return uint8(FUEL_BASE_ASSET_DECIMALS);
+    }
+
+    ////////////////////////
+    // Internal Functions //
+    ////////////////////////
+
+    /// @notice Performs all necessary logic to send a message to a target on Fuel
+    /// @param recipient The message receiver address or predicate root
+    /// @param data The message data to be sent to the receiver
+    function _sendOutgoingMessage(bytes32 recipient, bytes memory data) private {
+        bytes32 sender = bytes32(uint256(uint160(msg.sender)));
+        unchecked {
+            //make sure data size is not too large
+            require(data.length < MAX_MESSAGE_DATA_SIZE, "message-data-too-large");
+
+            //make sure amount fits into the Fuel base asset decimal level
+            uint256 precision = 10**(ETH_DECIMALS - FUEL_BASE_ASSET_DECIMALS);
+            uint256 amount = msg.value / precision;
+            if (msg.value > 0) {
+                require(amount * precision == msg.value, "amount-precision-incompatability");
+                require(amount <= ((2**64) - 1), "amount-precision-incompatability");
+            }
+
+            //emit message for Fuel clients to pickup (messageID calculated offchain)
+            emit SentMessage(sender, recipient, s_outgoingMessageNonce, uint64(amount), data);
+
+            // increment nonce for next message
+            ++s_outgoingMessageNonce;
+        }
+    }
+
+    /// @notice Executes a message in the given header
+    /// @param message The message to execute
+    /// @param blockHeader The block containing the message
+    /// @param messageInBlockProof Proof that message exists in block
+    function _executeMessageInHeader(
+        Message calldata message,
+        SidechainBlockHeader calldata blockHeader,
+        MerkleProof calldata messageInBlockProof
+    ) private {
+        //verify message validity
+        bytes32 messageId =
+            CryptographyLib.hash(
+                abi.encodePacked(
+                    message.sender,
+                    message.recipient,
+                    message.nonce,
+                    message.amount,
+                    message.data
+                )
+            );
+        require(!s_incomingMessageSuccessful[messageId], "Already relayed");
+        require(
+            (blockHeader.timestamp - 4611686018427387914) <=
```

Comment:
Magic numbers?

Reply:
Deserves a comment, but ref: https://unix.stackexchange.com/questions/283164/unix-seconds-tai-si-seconds-leap-seconds-and-real-world-code (
As one example, libtai provides a conversion (tai_now.c:7) to TAI from UXT as: TAI := 4,611,686,018,427,387,914 + UXT.)